bpf/bpf_host.c

```diff
@@ -1030,22 +1032,48 @@ static __always_inline int handle_l2_announcement(struct __ctx_buff *ctx)
 	if (ktime_get_ns() - (time) > L2_ANNOUNCEMENTS_MAX_LIVENESS)
 		return CTX_ACT_OK;
 
-	if (!arp_validate(ctx, &mac, &smac, &sip, &tip))
-		return CTX_ACT_OK;
+	if (!ip6) {
+		struct l2_responder_v4_key key;
 
-	key.ip4 = tip;
-	key.ifindex = ctx->ingress_ifindex;
-	stats = map_lookup_elem(&cilium_l2_responder_v4, &key);
-	if (!stats)
-		return CTX_ACT_OK;
+		if (!arp_validate(ctx, &mac, &smac, &sip, &tip))
+			return CTX_ACT_OK;
+
+		key.ip4 = tip;
+		key.ifindex = ctx->ingress_ifindex;
+		stats = map_lookup_elem(&cilium_l2_responder_v4, &key);
+		if (!stats)
+			return CTX_ACT_OK;
+
+		ret = arp_respond(ctx, &mac, tip, &smac, sip, 0);
+	} else {
+#ifdef ENABLE_IPV6
+		struct l2_responder_v6_key key6;
+		int l3_off;
+
+		if (!icmp6_ndisc_validate(ctx, ip6, &mac, &smac, &sip6, &tip6))
```

Comment:
`smac` and `sip6` are unused.

```suggestion
		if (!icmp6_ndisc_validate(ctx, ip6, &dmac, &tip6))
```

Reply:
`mac` and `sip`  are very much used, and are output params of the function: https://github.com/cilium/cilium/blob/369f79124da775f66c45f406da0c4c2c2ec53b67/bpf/lib/icmp6.h#L554-L557

I will fix the incorrect `const` in `smac` (cast in `eth_load_saddr()` masks this error).

Reply:
Damn.. this seems a bug in cocci:

https://github.com/cilium/cilium/actions/runs/16191362202/job/45707705664?pr=39648#step:4:15

Reply:
(Reverted to add the - incorrect - `const` statement for `smac`)

Reply:
Am I missing something? `handle_l2_announcement()` does not return `smac` or `sip6`. Yes, `icmp6_ndisc_validate()` returns these as output parameters. For the `arp_validate()` case, the equivalent source address parameters are subsequently passed to `arp_respond()`. But for this new IPv6 case, these are stack variables that are passed to `icmp6_ndisc_validate()` which returns them as output parameters, then the output parameters are not used later in `handle_l2_announcement()`.

Reply:
<img width="795" height="1283" alt="image" src="https://github.com/user-attachments/assets/82f458d6-c07c-4202-a722-11f3a2b26dd3" />

Reply:
Oh! Sorry, I didn't look into the caller :sweat_smile: 

Ok, `smac` removed from the param list, and `sip6` removed from param and from stack :rocket: